### PR TITLE
Improve C++ example

### PR DESCRIPTION
**Pull request by AntoineJT**
I think we can keep this C++ example simple but remove some C++ bad practices like `using namespace std` or using a '\n' instead of a `std::endl`.

So, here is my contribution, I hope you will merge it and enjoy it.

### CODE REVIEW

**Comment by Richienb** [comment]
That makes it more complicated.

**Comment by SirLynix** [comment]
`using namespace std;` is a [well-known bad practice](https://stackoverflow.com/questions/1452721/why-is-using-namespace-std-considered-bad-practice).

Plus, introducing the using namespace idiom does complicate the code imo, using objects from their namespaces is basic in C++.

**Comment by Richienb** [comment]
@SirLynix Ever heard of the fairytale The Man, The Boy and the Donkey? As soon as I accept this PR, another one will be opened within days changing it back. There are no winning moves.

**Comment by SirLynix** [comment]
I disagree, I gave you a link to explain why this is a bad thing. If someone opens a PR to put in back, send them this link.

Also, even you're right, isn't the point of this repository to have idiomatic Hello World code? As linked in my previous message, the current situation isn't idiomatic.

**Comment by AntoineJT** [comment]
@Richienb Please have in mind that the official C++ FAQ mentions this [here](https://isocpp.org/wiki/faq/coding-standards#using-namespace-std).

If someone else opens a PR to revert the changes, then just close the PR.
If the goal of this repository is to have bad practices hello world code, I can't understand its purpose.

I disagree with your argument also because that's an hello world repository. 
It makes sense to tell that some unexperimented people will try to change some hello world to make them something they will find "easier" or "better" but is in fact bad.
You can't do nothing about it else than closing the PRs. That should not affect the quality of the code that can be found in this repository.
[Here](https://github.com/leachim6/hello-world/pull/770) is a recent example of what I mean.
If I take this example, that's because I was looking at the network tab of this repository and then found this issue. I then look at the profile of that guy and [open an issue](https://github.com/thesmartdeveloperr/TicTacToeGame/issues/1) to help him write better code.

Here, I change the code because it contains bad practices, not to make it look as I wish.

**Comment by Richienb** [comment]
Alright then.

**Comment by Au-lit** [comment]
https://llvm.org/docs/CodingStandards.html#avoid-std-endl

**Comment by SirLynix** [comment]
It says "avoid std::endl;` not to say "use endl;" but to say "use \n", because std::endl; forces a flush.

`using namespace std;` is a recognized bad practice: https://stackoverflow.com/questions/1452721/why-is-using-namespace-std-considered-bad-practice

**Comment by Au-lit** [comment]
Yes, I was talking about using `'\n'` because ISO C++ flushes the `std::cout` stream at program return.

`using namespace std;` has no reason for being problematic for instance because we're in an implementation file and we're only making use of std things (but as I said; here it just adds some verbose).

**Comment by SirLynix** [comment]
This repository is about idiomatic code, best examples and such. ‘using namespace std;‘ is not one of those, as it can be problematic, even in implementations files.

What an idiomatic code would do to prevent the repetition of std:: (when using a lot of std::cout or std::endl) would be to use ‘using std::cout ;‘ locally to the function.

**Comment by Au-lit** [comment]
> This repository is about idiomatic code, best examples and such. using namespace std; is not one of those, as it can be problematic, even in implementations files.

That's what I said...

**Comment by Au-lit** [comment]
and even in this case `std::endl` is still bad practice

**Comment by Richienb** [comment]
It is possible to find evidence that all of the following are good practices (and the contrary):

- `std::cout`
- `std::endl`
- `\n`
- `using namespace std;`

I'm using the current method because of the clear backlash of rejecting it: https://github.com/leachim6/hello-world/pull/774#issuecomment-661308350

> and even in this case std::endl is still bad practice

If you can get at least 6 other people to like your comment (subtracting dislikes), I'll reconsider.

Keep in mind, I don't actually use C++ so I can only go with the majority.

**Comment by AllanBlanchard** [comment]
The C++ Core Guidelines (hosted by the ISO C++ website), tends to discourage the use of `std::endl`:

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rio-endl

For `using nnamespace std`, we found something similar in the FAQ:

https://isocpp.org/wiki/faq/coding-standards#using-namespace-std

(But they also sometimes do not follow their own rules, one example in the C++ Core Guidelines uses it).

**Comment by AntoineJT** [comment]
@Au-lit The LLVM doc you linked says "most of the time we don't need to flush the buffer" and that's true.
But here we only print one line and semantically it's better to keep std::endl and not '\n'.
This is not a bad practice because of a flush at the end of the program or whatever.
Flushing the buffer in a for loop when it's not needed is a bad practice because it is a slowdown in this case, but here we don't care, it is not a slowdown + semantically that's better : here we want to print a whole line at once...
We don't want to rely on a behavior triggered when the program is closed...

**Comment by SirLynix** [comment]
There's enough links on this PR to be clear about `using namespace std;`, which is a bad practice and shouldn't be used most of the time (and I think we all agree it shouldn't be present in an idiomatic hello world).

About \n vs std::endl honestly I'm not sure. `std::endl;` is a bad practice *when outputting multiple lines* indeed (where we would only want to flush once). But since here we're only outputting one line it doesn't matter much.
I'd prefer to keep the `std::endl` because that's what most people expect the most from a hello world, but I wouldn't be bothered if it was replaced by a \n.

**Comment by Richienb** [comment]
As someone just starting out learning C++, how should we teach them to make newline spaces?

**Comment by AntoineJT** [comment]
I will go for `sdl::endl`, as long as it flushes the buffer and so prints what the user wants as expected.
If for some reason he makes an infinite loop in which he prints some things but making newline with `\n` it will not have any output, and he will search for a while before to understand what's going on.
On the other hand, with `std::endl` it will have a performance issue compared to `\n`, but he will not search for a "bug" for a while, and its program will act as expected.
Following his learning of the language he will understand the difference between `\n` and `std::endl` without losing time by searching for a "bug".

**Comment by Au-lit** [comment]
In a `std::cout` with `\n` it will flush basically always.
So no he can't search for bugs with `\n` because they won't happend normally.

**Comment by Au-lit** [comment]
We just need to choose between one of those...
```cpp
#include <iostream>

int main() {
    std::cout << "Hello, new world!\n" << std::flush;
    //or
    std::cout << "Hello, new world!" << std::endl;
    //or
    std::cout << "Hello, new world!\n";
    // Done implicitly:
    return 0;
    std::cout.flush();
}
```

**Comment by AntoineJT** [comment]
Why are you showing some ugly code snippets as examples if you say "please don't pick that"? That makes no sense.
We don't have to choose anything anyway, because we can just keep the code as it is now, with a `std::endl`.
As I already said [here](https://github.com/leachim6/hello-world/pull/774#issuecomment-765344445), `std::endl` is better in this case imo because that's what a user expects when using `std::cout`.
If this
```quote
In a std::cout with \n it will flush basically always.
```
is true, then why using `\n` would be better for performances than a `std::endl` or just what the difference is between `std::endl` and `\n` except the fact that `\n` would have an error-prone behavior of "flushing or not" the buffer.
In this case, wouldn't `\n` be a bad practice because of a "I do whatever I want" and platform-specific behavior? (because that would be dependent of what the OS does when encountering a `\n`)
Then, why choose `\n` over `std::endl` here?
That's an hello world, and yes, we want to flush the buffer, because beginners don't know the notion of buffer or what it does. They just want to print something, to see it in the console.
Why bother them with a "maybe I will flush maybe not" behavior that they can encounter when trying to print in an endless program? (which will just be closed with a CTRL-C for example)

Even if you're right for the `std::cout` case, it's still a usage of streams. Streams are used also to write to files, and in an endless program like a "dump tool" for example, in order to debug a program (like I do in a private repository for a chip8 emulator) if you use `\n` without flushing the buffer, you'll get no content at all in the file when you close it with a CTRL-C afaik.
So it's still a pitfall for beginners.

Now, I want to know, why do you want to change this hello world?
It's a great hello world now, not a bad thing like before.

We don't want to explicitly flush the buffer with `std::flush` because iirc in this repository, hello world codes must be as simple as they can with good code.
If we explicitly flush the buffer, to make a newline we must use `\n` before the call to `std::flush`: it will literally be a `std::endl` but unnecessarily more verbose.

I think this would be better to lock this thread, in order to avoid some people wanting to change the code again, again and again for bad reasons.

**Comment by SirLynix** [comment]
While I agree that `endl` does flush one more time than required, I think it's better to leave it for a beginner.

This is a hello world example, we want to keep it idiomatic and simple. Switching from a commonly seen `endl` to a less commonly seen `\n` should be done for a good reason, and performance is not one of them. Remember that I/O is slow, like really really slow in C++.  
If performance was really a factor here, we should also use `std::cout.sync_with_stdio(false);` since by default **std::cout is unbuffered** anyways.  

TL;DR: I think that keeping `std::endl;` will be better for beginners, as performance isn't an issue for I/O for them. I do agree it's a bad practice but I think that performance shouldn't be a factor in a hello world example.
